lib/portage/dep/__init__.py

```diff
@@ -1562,6 +1600,7 @@ def __init__(
                 atom_re = _get_atom_wildcard_re(eapi_attrs)
                 m = atom_re.match(s)
                 if m is None:
+                    _atom_cache[s] = InvalidAtom(self)
```

Comment:
Is there a real gain from caching invalid atoms? I mean, they shouldn't occur during normal operation, should they?

Reply:
There's lots of different places that instantiate an `Atom`, and while I didn't check them all super thoroughly, I think some of them use logic around whether or not they were able to succesfully create one.

Similarly, my understanding is that in python, it is idiomatic to use exceptions for control flow.

That's why I chose to cache the exception, so that I could throw it sooner rather than having to plough through the huge `__init__` just to get to it again.